modules/misc/nixgl.nix

```diff
@@ -1,29 +1,166 @@
+        shorthand.
+
+        The following wrappers are available:
+        ${wrapperListMarkdown}
+      '';
+    };
+
+    prime.card = lib.mkOption {
+      type = lib.types.str;
+      default = "1";
+      example = "pci-0000_06_00_0";
+      description = ''
+        Selects the non-default graphics card used for PRIME render offloading.
+        The value can be:
+
+        - a number, selecting the n-th non-default GPU;
+        - a PCI bus id in the form `pci-XXX_YY_ZZ_U`;
+        - a PCI id in the form `vendor_id:device_id`
+
+        For more information, consult the Mesa documentation on the `DRI_PRIME`
+        environment variable.
+      '';
+    };
+
+    prime.nvidiaProvider = lib.mkOption {
+      type = with lib.types; nullOr str;
+      default = null;
+      example = "NVIDIA-G0";
+      description = ''
+        If this option is set, it overrides the offload provider for Nvidia
+        PRIME offloading. Consult the proprietary Nvidia driver documentation
+        on the `__NV_PRIME_RENDER_OFFLOAD_PROVIDER` environment variable.
+      '';
+    };
+
+    prime.installScript = lib.mkOption {
+      type = with lib.types; nullOr (enum [ "mesa" "nvidia" ]);
+      default = null;
+      example = "mesa";
+      description = ''
+        If this option is set, the wrapper script `prime-offload` is installed
+        into the environment. It allows starting programs on the secondary GPU
+        selected by the `nixGL.prime.card` option. This makes sense when the
+        program is not already using one of nixGL PRIME wrappers, or for
+        programs not installed from Nixpkgs.
+
+        This option can be set to either "mesa" or "nvidia", making the script
+        use one or the other graphics library.
+      '';
+    };
+
+    installScripts = lib.mkOption {
+      type = with lib.types;
+        nullOr (listOf (enum (builtins.attrNames config.lib.nixGL.wrappers)));
+      default = null;
+      example = [ "mesa" "mesaPrime" ];
+      description = ''
+        For each wrapper `wrp` named in the provided list, a wrapper script
+        named `nixGLWrp` is installed into the environment. These scripts are
+        useful for running programs not installed via Home Manager.
+
+        The following wrappers are available:
+        ${wrapperListMarkdown}
+      '';
+    };
+
+    vulkan.enable = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      example = true;
+      description = ''
+        Whether to enable Vulkan in nixGL wrappers.
+
+        This is disabled by default bacause Vulkan brings in several libraries
+        that can cause symbol version conflicts in wrapped programs. Your
+        mileage may vary.
+      '';
+    };
   };
 
-  config = {
-    lib.nixGL.wrap = # Wrap a single package with the configured nixGL wrapper
-      pkg:
+  config = let
+    getWrapperExe = vendor:
+      let
+        glPackage = cfg.packages.${pkgs.system}."nixGL${vendor}";
```

Comment:
Will this work for non-flake setups? The manual claims the packages can be provided from a channel, but nixGL's `default.nix` doesn't seem to have this structure. 

I think this is my only remaining concern for the interface, that it couples this module to the internal structure of nixGL, which may change in the future. I wonder if we should keep the prefix-style arguments (`prefix`, `offloadPrefix`, and `vulkanPackage`) to keep this module decoupled from nixGL's internal structure.

Reply:
Is non flake support a requirement for all hm integrations? Or could that support be added after the initial merge?

Reply:
And re: coupling to the internal structure of nixGL
the wrappers do seem useful, would adding an assertion in the wrapper about the expected structure of nixGL be a terrible idea?

Reply:
You're right, of course; the structure of nixGL is convoluted. However, I see this HM module as a way to shield the user from this complexity, to a _resonable_ extent. When I first tried to integrate nixGL into my HM config, I spent a fair bit of time banging my head against the desk because stuff was documented as being under `auto`, but wasn't because I was using flakes. I'd like to spare people having to deal with that.

I think there is a way to deal with the difference in structure between channels and flakes in such a way that, should the attributes ever be moved from `auto` to the level above (which would make sense), the module won't break. However, I was unable to develop a solution because I hit a strange issue.

My own HM setup is a flake, so I wanted to spin up a VM to play with channels. So, I made a clean install of Debian (also tried Fedora), a clean setup of Nix using the "`curl | sh`" method, and found that nixGL doesn't build. For example, this command

```sh
NIXPKGS_ALLOW_UNFREE=1 nix --extra-experimental-features nix-command --extra-experimental-features flakes run --impure github:nix-community/nixGL#nixGLNvidia
```

yields the error 

```
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:34:12:
           33|
           34|   strict = derivationStrict drvAttrs;
             |            ^
           35|

       … while evaluating the derivation attribute 'name'
         at /nix/store/kc7nh9sdr4z8c7xc9j9891v212k2zj8k-source/pkgs/stdenv/generic/make-derivation.nix:270:7:
          269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          270|       name =
             |       ^
          271|         let

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: cannot coerce null to a string: null
```

Similarly, trying to use channels by doing

```sh
nix-channel --add https://github.com/nix-community/nixGL/archive/main.tar.gz nixgl && nix-channel --update
nix-env -iA nixgl.auto.nixGLNvidia
```

results in

```
error:
       … in the condition of the assert statement
         at /nix/store/z7dky1g7nnxmfq7x6mz6xxswkyxmrflx-nixpkgs/nixpkgs/pkgs/build-support/trivial-builders/default.nix:89:5:
           88|     }:
           89|     assert lib.assertMsg (destination != "" -> (lib.hasPrefix "/" destination && destination != "/")) ''
             |     ^
           90|       destination must be an absolute path, relative to the derivation's out path,

       … in the left operand of the OR (||) operator
         at /nix/store/z7dky1g7nnxmfq7x6mz6xxswkyxmrflx-nixpkgs/nixpkgs/lib/asserts.nix:42:10:
           41|     msg:
           42|     pred || builtins.throw msg;
             |          ^
           43|

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: cannot coerce null to a string: null
```

I have no clue what's going on. It works fine on my system. Any ideas?

Reply:
I've found I often need `--show-trace` to make sense of the error, otherwise the important parts get removed

I could try to recreate/debug, but want to follow the same steps for VM creation, do you have a link to a guide or something?

Reply:
I think both y'all @exzombie and @Smona know far more about this than I do, but I'm here for (trying to) help. I used to use non-flake setup, can try to restore that, but slightly worried about the "re-efficacy" of my setup w/o some guidelines on a VM

Reply:
I literally only installed a fresh Debian into a VM, did a multiuser install of nix, rebooted, then tried the above commands. They really should work on a clean install, methinks. I think I'm going crazy, so I would appreciate if you could at least reproduce the issue.

And yes, `--show-trace` is often useful, if verbose. Didn't help me any in this particular case.

Reply:
I am working on trying to reproduce that issue, but running into some issues getting the virtual machine running, will update when I'm able to test it out

Reply:
As much as i hate that flakes are still considered experimental, and wish for them to be standardized soon, there are still some niche outstanding issues that might prevent someone from wanting to migrate to flakes yet ([lazy trees](https://github.com/NixOS/nix/issues/3121) come to mind), and ideally this module would support those users.

It looks like that second error is coming from `writeTextFile`, which is only used in `nixGL` for [writeExecutable](https://github.com/nix-community/nixGL/blob/310f8e49a149e4c9ea52f1adf70cdc768ec53f8a/nixGL.nix#L22). It suggests to me that a null name is somehow being passed there, but i don't see how that's possible. It might be coming from deeper in the stack.

My initial instinct is to write this off as "not worth root causing". There may be some limitations to your VM environment that prevent nixGL from running as expected. Could you instead test the non-flake version inside your flake config on bare metal, via `builtins.fetchGit`? I'm imagining something like this should work:

```nix
let nixGLSrc = builtins.fetchGit {
  url = "git@github.com:nix-community/nixGL.git";
  ref = "main";
  rev = "310f8e49a149e4c9ea52f1adf70cdc768ec53f8a";
};
nixGL = import "${nixGLSrc}/default.nx" { 
  pkgs = inputs.nixpkgs.legacyPackages.${system}; 
  # other configurations...
};
```

Reply:
For testing the non-flake setup on bare-metal. I think having an existing flake install shouldn't matter, right? If that's not an issue I'll also test it out

Reply:
> For testing the non-flake setup on bare-metal. I think having an existing flake install shouldn't matter, right? If that's not an issue I'll also test it out

I wish. I made a small home configuration, and got more weird evaluation errors. Searching shows similar errors occurred when people didn't pin their nixpgks properly. I must admit I do not have much interest in pursuing this further, seeing as I don't use channels.

@Smona, your "faux-channel" suggestion worked perfectly. I pushed a change which works both with a flake and with `default.nix`, and also extended the documentation with what I assume things would look like if one were to use an actual channel.